### PR TITLE
[LLM Runtime] Streaming-LLM based on shift RoPE

### PR DESCRIPTION
**Pull request by DDEle**
## Type of Change: feature
API change: add switch `--shift-roped-k` to the C++ API 

## Description
Support shifting RoPE-ed K-cache after the KV-cache is filled up **for llama**.

Per-model changes:
- `n_ctx` is now a field of `model_context` rathe then model class themselves, together with the added `n_keep` and `shift_roped_k`. It is also more semantic as `n_ctx` is read from args rather than model files. 
- Added an extra argument `n_total` for `xxx_model_eval_internal`, which provides the number of tokens evaluated so far (including evicted tokens if there is any). Note: `n_past` is used as the offset to which the KV is cached to even before this PR
- in `xxx_utils.cpp`, the template argument of the `unique_ptr` in `model_load_internal` are corrected to the derived classes
- Time are calculated outside of the `xxx_utils.cpp`

Note: The version for fused-attention will be implemented in a later PR

## Expected Behavior & Potential Risk
N/A

## How has this PR been tested?
<details>
<summary>Accuracy Tests Details:</summary>

- <details>
  <summary>
  graph-streaming-llm-shift-rope <code>numactl -m 1 -C 56-111 bin/run_llama -m /home_local/dingyi/data/llama-7b-hf-pr573-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 1024 -c 64 -n 128 --keep 4 --memory-f16 --shift-roped-k -p "$GIRL_PROMPT"</code>
  </summary>

  > Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun.
Then one day, her parents took the little girl on an adventure, that she would never forget. They drove their tiny car, through a dark tunnel, where nothing could be seen, but the stars above. Afterwards, as the days went by, the little girl thought of this place, and how she felt there, inside of it. It was not a place to get lost, or to find treasure. The child wondered at its vastness, with so many people there, and yet no one ever spoke, or even moved very much, except those who were in service.
She did know, though,
  </details>
</details>

## Dependency Change?
No

### CODE REVIEW

**Review by a32543254 — APPROVED** [review]
LGTM

**Comment by DDEle** [comment]
https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/159

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c [hunk 1]
```diff
@@ -7660,14 +7667,17 @@ static void ne_compute_forward_rope_f32(const struct ne_compute_params* params,
   if (params->type == NE_TASK_INIT || params->type == NE_TASK_FINALIZE) {
     return;
   }
+  NE_ASSERT(src1->type == NE_TYPE_I32);
+  NE_ASSERT(ne_nelements(src1) == 5);  // 5 params
 
-  float freq_base = 10000.0f;
-  float freq_scale = 1.0f;
+  const float freq_base = 10000.0f;
+  const float freq_scale = 1.0f;
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:7674]: static

**Review by airMeng — APPROVED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 2]
```diff
@@ -296,6 +301,9 @@ model_token Model::post_process(float* logits) {
     } else {
       return post_sample_top_k_top_p_repeat(logits);
     }
+  } else {
+    fprintf(stderr, "error, beam search not supported!\n");
```
Comment (zhenwei-intel) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:305]: beam search supported in generate_tokens

**Review by zhenwei-intel — APPROVED** [review]

**Review by zhentaoyu — APPROVED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 2 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp]: updated in fdf5a88

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c]: updated in fdf5a88

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c [hunk 3]
```diff
@@ -7830,11 +7844,55 @@ static void ne_compute_forward_rope_f16(const struct ne_compute_params* params,
...
+          if (ir++ < ir0) continue;
+          if (ir > ir1) break;
+          if (i2 < n_keep) continue;  // The "break" will break `ir`
+
+          for (int64_t i0 = 0; i0 < ne0; i0 += 2) {  // along head size
+            const ne_fp16_t* const src =
+                (ne_fp16_t*)((char*)src0->data + i3 * nb03 + i2 * nb02 + i1 * nb01 + i0 * nb00);
+            ne_fp16_t* dst_data = (ne_fp16_t*)((char*)dst->data + i3 * nb3 + i2 * nb2 + i1 * nb1 + i0 * nb0);
+
+            const float x0 = NE_FP16_TO_FP32(src[0]);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:7878]: can we vectorize here? @yuchengliu1 in case

**Comment by VincyZhang** [comment]
update extension test: https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension/164/artifact/report.html

File: intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c (hunk 3 above)
Comment (yuchengliu1) [reply intel_extension_for_transformers/llm/runtime/graph/core/ne_layers.c:7878]: There is no vectorized FP16 to FP32 convert now. We can add it if we need.
